introduction_to_amazon_algorithms/jumpstart-foundation-models/text-generation-benchmarking/benchmarking/load_test.py

```diff
@@ -0,0 +1,344 @@
+
+    def run_latency_load_test(
+        self,
+        num_invocations: int,
+        retry_wait_time: float = RETRY_WAIT_TIME_SECONDS,
+        max_total_retry_time: float = MAX_TOTAL_RETRY_TIME_SECONDS,
+    ) -> Dict[str, Any]:
+        logging.info(f"{self._logging_prefix} Begin latency load test ...")
+        time.sleep(CLOUDWATCH_PERIOD)  # wait for 1 cloudwatch period to ensure no extra queries are reported
+        statistics_latency = self.run_load_test(num_invocations, 1)
+        metrics = self._extract_cloudwatch_metrics(statistics_latency, retry_wait_time, max_total_retry_time)
+        metrics["Client"] = statistics_latency.get_statistics(self.tokenizer, self.price_per_endpoint)
+        metrics["ModelID"] = self.model_id
+        metrics["PayloadName"] = self.payload_name
+        return metrics
+
+    def run_throughput_load_test(self, num_invocations: int, max_workers: int) -> Dict[str, Any]:
+        logging.info(f"{logging_prefix(self.model_id, self.payload_name, max_workers)} Begin throughput load test ...")
+        statistics_throughput = self.run_load_test(num_invocations, max_workers)
+        metrics = statistics_throughput.get_statistics(self.tokenizer, self.price_per_endpoint)
+        metrics.update(
+            {
+                "ModelID": self.model_id,
+                "PayloadName": self.payload_name,
+                "Invocations": num_invocations,
+                "ConcurrentRequests": max_workers,
+            }
+        )
+        return metrics
+
+    def run_concurrency_probe(
+        self,
+        iterator_cls: Type[ConcurrentProbeIteratorBase],
+        num_invocation_hook: Callable[[int], int],
+    ) -> List[Dict[str, Any]]:
+        """Probe the endpoint with a series of concurrent request payloads to obtain a set of throughput measures.
+
+        Arguments:
+            concurrent_request_iterator (ConcurrentRequestIteratorBase): An iterator that controls the number of
+                concurrent requests to load the endpoint with during the probe.
+            num_invocation_hook (Callable[[int], int]): A hook that controls the number of invocations use during a
+                single load test as a function of the number of concurrent requests.
+        """
+        logging.info(f"{self._logging_prefix} Begin concurrency probe ...")
+        results: List[Dict[str, Any]] = []
+        concurrent_request_iterator = iterator_cls(self.model_id, self.payload_name)
+        for concurrent_requests in concurrent_request_iterator:
+            try:
+                num_invocations = num_invocation_hook(concurrent_requests)
+                result = self.run_throughput_load_test(num_invocations, concurrent_requests)
+            except Exception as e:
+                concurrent_request_iterator.exception = e
+            else:
+                if concurrent_request_iterator.send(result, self.predictor):
```

Comment:
question: why not move this line to inside the `try` block? Or is that because you want to throw if iterating fails?

Reply:
I'll try with this in `try` -- I don't think it matters. Importantly, the exception is set on the iterator and the iterator can determine whether or not to raise a `StopIteration` based on this exception. Interesting potential use case: imagine an iterator that runs until failure, but instead of raising `StopIteration`, it scales the endpoint to more instances, blocks during a wait period, and then retries the iteration. This behavior can be written in a custom iterator outside of the core logic here.